src/game/movement.rs

```diff
@@ -120,36 +119,122 @@ fn update_facing(mut player_query: Query<(&MovementController, &mut Sprite)>) {
+
+        let animation_state = if controller.0 == Vec2::ZERO {
+            PlayerAnimationState::Idle
+        } else {
+            PlayerAnimationState::Walking
+        };
+        animation.update_state(animation_state);
+
+        if animation.changed() {
+            atlas.index = animation.get_atlas_index();
+        }
+    }
+}
+
+/// If the player is moving, play a step sound effect synchronized with the animation.
+fn trigger_step_sfx(mut commands: Commands, mut step_query: Query<&PlayerAnimation>) {
+    for animation in &mut step_query {
+        if animation.state == PlayerAnimationState::Walking
+            && animation.changed()
+            && (animation.frame == 2 || animation.frame == 5)
+        {
+            commands.trigger(Sfx::Step);
+        }
+    }
+}
+
+/// Component that tracks player's animation state.
+/// It is tightly bound to the texture atlas we use.
 #[derive(Component, Reflect)]
 #[reflect(Component)]
-pub struct StepSfx {
-    pub cooldown_timer: Timer,
+pub struct PlayerAnimation {
+    timer: Timer,
+    frame: usize,
+    state: PlayerAnimationState,
 }
 
-impl StepSfx {
-    pub fn new(cooldown: Duration) -> Self {
-        let mut cooldown_timer = Timer::new(cooldown, TimerMode::Once);
-        cooldown_timer.set_elapsed(cooldown);
-        Self { cooldown_timer }
-    }
+#[derive(Reflect, PartialEq)]
+pub enum PlayerAnimationState {
+    Idle,
+    Walking,
 }
 
-fn tick_step_sfx(time: Res<Time>, mut step_query: Query<&mut StepSfx>) {
-    for mut step in &mut step_query {
-        step.cooldown_timer.tick(time.delta());
+impl PlayerAnimation {
+    /// How many idle frames are there.
+    const IDLE_FRAMES: usize = 2;
+    /// What's the interval between idle frames.
+    const IDLE_INTERVAL: Duration = Duration::from_millis(500);
+
+    fn idle() -> Self {
+        Self {
+            timer: Timer::new(Self::IDLE_INTERVAL, TimerMode::Repeating),
+            frame: 0,
+            state: PlayerAnimationState::Idle,
+        }
     }
-}
 
-/// If the player is moving, play a step sound effect.
-fn trigger_step_sfx(
-    mut commands: Commands,
-    mut step_query: Query<(&MovementController, &mut StepSfx)>,
-) {
-    for (controller, mut step) in &mut step_query {
-        if step.cooldown_timer.finished() && controller.0 != Vec2::ZERO {
-            step.cooldown_timer.reset();
-            commands.trigger(Sfx::Step);
+    /// How many walking frames are there.
+    const WALKING_FRAMES: usize = 6;
+    /// What's the interval between walking frames.
+    const WALKING_INTERVAL: Duration = Duration::from_millis(50);
+
+    fn walking() -> Self {
+        Self {
+            timer: Timer::new(Self::WALKING_INTERVAL, TimerMode::Repeating),
+            frame: 0,
+            state: PlayerAnimationState::Walking,
+        }
+    }
+
+    pub fn new() -> Self {
+        Self::idle()
+    }
+
+    /// Update animation timers.
+    pub fn update_timer(&mut self, delta: Duration) {
```

Comment:
Would name this `PlayerAnimation::tick` to match `Timer::tick`.

Reply:
This matches `update_state`, which is better since they're similar methods on the same object

Reply:
I'd argue that update animation should be split into two systems, one to tick the timer and one to update the animation, but this is fine I guess.

Reply:
They do very different things so I'm not personally convinced, but alright.

Reply:
Well I'm not convinced we should reflect timer API in animation controller, because it contains one :p